### DISCUSSION TITLE
运行于Termux上的问题 problems when running on Termux

### DISCUSSION THREAD

**Discussion by Octopustank** [dashingsoft/pyarmor#1371] [Q&A]
_Termux安装于HarmonyOS。Python版本1均在311_
### 1.使用``pyarmor -d gen test.py``
返回：

    INFO     Python 3.11.4 INFO
    Pyarmor 8.2.9 (trial), 000000, non-profits
    INFO     Platform linux.aarch64
    DEBUG    native platform linux.aarch64
    DEBUG    home path: /data/data/com.termux/files/home/.pyarmor
    DEBUG    command options: {'no_runtime': False, 'inputs': ['test.py']}
    DEBUG    install plugin: CodesignPlugin
    INFO     search inputs ...
    INFO     find script test.py
    INFO     find 1 top resources
    [1]    5030 segmentation fault  pyarmor -d gen test.py

test.py在Windows_x86_64下正常加密
### 2.打开在Windows_x86_64下加密，欲运用于Termux：
发现要指定andriod_aarch64而非linux_aarch64；
指定andriod_aarch64后，使用``python test.py``后返回：

    Traceback (most recent call last):
      File "/storage/emulated/0/py_files/dist/test.py", line 2, in <module>
        from pyarmor_runtime_000000 import __pyarmor__
      File "/storage/emulated/0/py_files/dist/pyarmor_runtime_000000/__init__.py", line 49, in <module>
        __pyarmor__ = __pyarmor__().__pyarmor__
                           ^^^^^^^^^^^^^
      File "/storage/emulated/0/py_files/dist/pyarmor_runtime_000000/__init__.py", line 48, in __pyarmor__
        return __import__(name, globals(), locals(), ['__pyarmor__'], level=1) 
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ImportError: dlopen failed: library "/storage/emulated/0/py_files/dist/pyarmor_runtime_000000/android_aarch64/pyarmor_runtime.so" needed or dlopened by "/data/data/com.termux/files/usr/lib/libpython3.11.so.1.0" is not accessible for the namespace "(default)"

如何解决呢？或者还应询问Termux开发者

**Comment by jondy** [comment]
权限问题，动态库存放在 `/storage/` 下面是无法装载的，需要把运行辅助库 `pyarmor_runtime_000000` 放到有权限的目录，具体请参考 Android 的文档。

**Reply by Octopustank** [reply]
我将其移到了/data/data/com.termux/files/home,不再报先前的Python错了，但报了segmentation fault。这个是可以解决的吗？

**Comment by jondy** [comment]
使用 `ldd /path/to/pytransform3.so` 查看是否缺少依赖的动态库。

**Reply by Octopustank** [reply]
该命令返回

        libdl.so => /system/lib64/libdl.so
        libm.so => /system/lib64/libm.so
        libc.so => /system/lib64/libc.so
        ld-android.so => /system/lib64/ld-android.so

经find查看/system/lib64下都有文件

**Comment by jondy** [comment]
What's the output of `ldd $(which python)`?

**Reply by Octopustank** [reply]
$ ldd $(which python)
        libandroid-support.so =>/data/data/com.termux/files/usr/lib/libandroid-support.so
        libpython3.11.so.1.0 =>/data/data/com.termux/files/usr/lib/libpython3.11.so.1.0
        libc.so => /system/lib64/libc.so
        ld-android.so => /system/lib64/ld-android.so
        libdl.so => /system/lib64/libdl.so
        libm.so => /system/lib64/libm.so

**Comment by jondy** [comment]
需要重新编译一下 pyarmor_runtime.so

**Reply by Octopustank** [reply]
是再次使用``pyarmor gen --platform android.aarch64  --platform windows.x86_64 test.py``吗（在windows上执行）？
Termux上，该命令一定会segmentation fault，就像Termux上运行加密后的py文件一样。

**Reply by jondy** [reply]
不是的，需要 Pyarmor 开发组重新编译一下 pyarmor_runtime.so，在下一个发布版本里面。

**Reply by Octopustank** [reply]
那麻烦了～谢谢开发组

**Comment by Octopustank** [comment]
平台：安装于Android的Termux，使用pyarmor版本：Version: 8.3.1，使用pyarmor -d gen test.py，
仍报错
> DEBUG    unknown error, please check pyarmor.error.log
ERROR    dlopen failed: cannot locate symbol "PyFloat_Type" referenced by "/data/data/com.termux/files/usr/lib/python3.11/site-packages/pyarmor/cli/core/android/aarch64/pytransform3.so"...


在PC上进行生成，在放到安装于Android的Termux，仍报错

> Traceback (most recent call last):
  File "/data/data/com.termux/files/home/dist/test.py", line 2, in <module>
    from pyarmor_runtime_000000 import __pyarmor__
  File "/data/data/com.termux/files/home/dist/pyarmor_runtime_000000/__init__.py", line 49, in <module>
    __pyarmor__ = __pyarmor__().__pyarmor__
                  ^^^^^^^^^^^^^
  File "/data/data/com.termux/files/home/dist/pyarmor_runtime_000000/__init__.py", line 48, in __pyarmor__
    return __import__(name, globals(), locals(), ['__pyarmor__'], level=1)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ImportError: dlopen failed: cannot locate symbol "PyBytes_AsStringAndSize" referenced by "/data/data/com.termux/files/home/dist/pyarmor_runtime_000000/android_aarch64/pyarmor_runtime.so"...

**Comment by jondy** [comment]
加密脚本的时候使用的 Python 版本是 3.11 吗？

**Comment by Octopustank** [comment]
是的，Windows和termux上均为3.11.4



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2023年8月14日(周一) 晚上9:03
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [dashingsoft/pyarmor] 运行于Termux上的问题 problems when running on Termux (Discussion #1371)




 
加密脚本的时候使用的 Python 版本是 3.11 吗？

**Comment by jondy** [comment]
那么 `/data/data/com.termux/files/usr/lib/libpython3.11.so.1.0` 是否包含这些未定义的函数呢？
```
objdump -t /data/data/com.termux/files/usr/lib/libpython3.11.so.1.0 | grep PyFloat_Type
objdump -t /data/data/com.termux/files/usr/lib/libpython3.11.so.1.0 | grep PyBytes_AsStringAndSize
```

如果没有的话，必须重新编译 termux 的 Python, 允许输出这些函数。

扩展模块 `pyarmor_runtime.so` 和 `pytransform3.so` 需要 Python 动态库提供这些函数。

**Comment by jondy** [comment]
另外可以测试一下是否需要设置 `sys.setdlopenflags(os.RTLD_GLOBAL)`，在 termux 下面运行
```
python3.11
>>> import os, sys
>>> sys.setdlopenflags(os.RTLD_GLOBAL)
>>> import pyarmor.cli.core.pytransform3
```

**Comment by Octopustank** [comment]
运行结果：
Python 3.11.4 (main, Jul  2 2023, 11:17:00) [Clang 14.0.7 (https://android.googlesource.com/toolchain/llvm-project 4c603efb0 on linux
Type "help", "copyright", "credits" or "license" for more information.
&gt;&gt;&gt; import os, sys
&gt;&gt;&gt; sys.setdlopenflags(os.RTLD_GLOBAL)
&gt;&gt;&gt; import pyarmor.cli.core.pytransform3
Traceback (most recent call last):
&nbsp; File "<stdin&gt;", line 1, in <module&gt;
ModuleNotFoundError: No module named 'pyarmor.cli.core.pytransform3'



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2023年8月15日(周二) 中午12:27
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [dashingsoft/pyarmor] 运行于Termux上的问题 problems when running on Termux (Discussion #1371)




 
另外可以测试一下是否需要设置 sys.setdlopenflags(os.RTLD_GLOBAL)，在 termux 下面运行
 python3.11 &gt;&gt;&gt; import os, sys &gt;&gt;&gt; sys.setdlopenflags(os.RTLD_GLOBAL) &gt;&gt;&gt; import pyarmor.cli.core.pytransform3

**Comment by Octopustank** [comment]
两条均进行运行后，在很长一段时间后均未结束
1：
$ objdump -t /data/data/com.termux/files/usr/lib/libpython3.11.so.1.0 &amp; grep PyFloat_Type
[1] 11681


/data/data/com.termux/files/usr/lib/libpython3.11.so.1.0:    file format elf64-littleaarch64


SYMBOL TABLE:
[1]  + 11681 done       objdump -t /data/data/com.termux/files/usr/lib/libpython3.11.so.1.0
（未结束）


2：
$ objdump -t /data/data/com.termux/files/usr/lib/libpython3.11.so.1.0 &amp; grep PyBytes_AsStringAndSize
[1] 10212


/data/data/com.termux/files/usr/lib/libpython3.11.so.1.0:    file format elf64-littleaarch64


SYMBOL TABLE:
[1]  + 10212 done       objdump -t /data/data/com.termux/files/usr/lib/libpython3.11.so.1.0
（未结束）



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2023年8月15日(周二) 中午12:10
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [dashingsoft/pyarmor] 运行于Termux上的问题 problems when running on Termux (Discussion #1371)




 
那么 /data/data/com.termux/files/usr/lib/libpython3.11.so.1.0 是否包含这些未定义的函数呢？
 objdump -t /data/data/com.termux/files/usr/lib/libpython3.11.so.1.0 | grep PyFloat_Type objdump -t /data/data/com.termux/files/usr/lib/libpython3.11.so.1.0 | grep PyBytes_AsStringAndSize  
如果没有的话，必须重新编译 termux 的 Python, 允许输出这些函数。
 
扩展模块 pyarmor_runtime.so 和 pytransform3.so 需要 Python 动态库提供这些函数。

**Comment by jondy** [comment]
那么这个呢

```
>>> import os, sys
>>> sys.setdlopenflags(os.RTLD_GLOBAL)
>>> import pyarmor.cli.core.android.aarch64.pytransform3
```

**Comment by Octopustank** [comment]
&gt;&gt;&gt; import os, sys
&gt;&gt;&gt; sys.setdlopenflags(os.RTLD_GLOBAL)
&gt;&gt;&gt; import pyarmor.cli.core.android.aarch64.pytransform3
Traceback (most recent call last):
&nbsp; File "<stdin&gt;", line 1, in <module&gt;
ImportError: dlopen failed: cannot locate symbol "PyFloat_Type" referenced by "/data/data/com.termux/files/usr/lib/python3.11/site-packages/pyarmor/cli/core/android/aarch64/pytransform3.so"...



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2023年8月15日(周二) 下午4:48
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [dashingsoft/pyarmor] 运行于Termux上的问题 problems when running on Termux (Discussion #1371)




 
那么这个呢
 &gt;&gt;&gt; import os, sys &gt;&gt;&gt; sys.setdlopenflags(os.RTLD_GLOBAL) &gt;&gt;&gt; import pyarmor.cli.core.android.aarch64.pytransform3

**Comment by jondy** [comment]
I don't know why in termux `pytransform3.so` could not get symbol `PyFloat_Type` from `libpython3.11.so.1.0`, it should be exported from `libpython3.11.so.1.0`.

**Reply by Octopustank** [reply]
May be pull an issue on https://github.com/termux/termux-app ?
